Services/Tree/classes/Setup/class.ilTreeImplementationSwitch.php

```diff
@@ -0,0 +1,147 @@
+{
+    public const NESTED_SET_TO_MATERIALIZED_PATH = 'nested_set_to_materialized_path';
+    public const MATERIALIZED_PATH_TO_NESTED_SET = 'materialized_path_to_nested_set';
+
+    private string $mode;
+
+    public function __construct(string $mode)
+    {
+        parent::__construct(new NullConfig());
+        $this->mode = $mode;
+    }
+
+    public function getHash() : string
+    {
+        return hash('sha256', self::class);
+    }
+
+    public function getLabel() : string
+    {
+        if ($this->mode === self::MATERIALIZED_PATH_TO_NESTED_SET) {
+            return 'The tree implementation is switched to nested set';
+        }
+
+        return 'The tree implementation is switched to materialized path';
+    }
+
+    public function isNotable() : bool
+    {
+        return true;
+    }
+
+    public function getPreconditions(Environment $environment) : array
+    {
+        return [
+            new ilSettingsFactoryExistsObjective(),
+            new ilDatabaseInitializedObjective(),
+        ];
+    }
+
+    public function achieve(Environment $environment) : Environment
+    {
+        $db = $environment->getResource(Environment::RESOURCE_DATABASE);
+        $settings_factory = $environment->getResource(Environment::RESOURCE_SETTINGS_FACTORY);
+
+        $is_nested_set = $settings_factory->settingsFor('common')->get('main_tree_impl', 'ns') === 'ns';
+
+        if ($this->mode === self::NESTED_SET_TO_MATERIALIZED_PATH && $is_nested_set) {
+            ilMaterializedPathTree::createFromParentRelation($db);
+
+            $db->dropIndexByFields('tree', ['lft']);
+            $db->dropIndexByFields('tree', ['path']);
+            $db->addIndex('tree', ['path'], 'i4');
+
+            $settings_factory->settingsFor('common')->set('main_tree_impl', 'mp');
+        } elseif ($this->mode === self::MATERIALIZED_PATH_TO_NESTED_SET && !$is_nested_set) {
+            $renumber_callable = function (ilDBInterface $db) {
+                $this->renumberNestedSet($db, 1, 1);
+            };
+            $ilAtomQuery = $db->buildAtomQuery();
+            $ilAtomQuery->addTableLock('tree');
+
+            $ilAtomQuery->addQueryCallable($renumber_callable);
+            $ilAtomQuery->run();
+
+            $db->dropIndexByFields('tree', ['lft']);
+            $db->dropIndexByFields('tree', ['path']);
+            $db->addIndex('tree', ['lft'], 'i4');
+
+            $settings_factory->settingsFor('common')->set('main_tree_impl', 'ns');
+        } else {
+            throw new UnachievableException('The tree implementation switch does already equal the requested mode');
+        }
+
+        return $environment;
+    }
+    
+    private function renumberNestedSet(ilDBInterface $db, int $node_id, int $i) : int
+    {
+        $db->manipulateF(
+            'UPDATE tree SET lft = %s WHERE child = %s',
+            [ilDBConstants::T_INTEGER, ilDBConstants::T_INTEGER],
+            [$node_id, $i]
```

Comment:
I've switched these parameter. The result was quite surprising :)

Reply:
Thanks, to both of you
I 've cherry-picked your commits to R8 and will close this PR.